File: aws-globalaccelerator-listener/src/main/java/software/amazon/globalaccelerator/listener/CreateHandler.java
package software.amazon.globalaccelerator.listener;

import com.amazonaws.services.globalaccelerator.AWSGlobalAccelerator;
import com.amazonaws.services.globalaccelerator.model.*;
import com.amazonaws.services.globalaccelerator.model.PortRange;
import lombok.val;
import software.amazon.cloudformation.proxy.*;

import java.util.stream.Collectors;

public class CreateHandler extends BaseHandler<CallbackContext> {
    @Override
    public ProgressEvent<ResourceModel, CallbackContext> handleRequest(
        final AmazonWebServicesClientProxy proxy,
        final ResourceHandlerRequest<ResourceModel> request,
        final CallbackContext callbackContext,
        final Logger logger
    ) {
        logger.log(String.format("Creating listener with request [%s]", request));

        val agaClient = AcceleratorClientBuilder.getClient();
        val inferredCallbackContext = callbackContext != null ?
                callbackContext :
                CallbackContext.builder()
                        .stabilizationRetriesRemaining(HandlerCommons.NUMBER_OF_STATE_POLL_RETRIES)
                        .build();

        // confirm we can find the accelerator
        final ResourceModel model = request.getDesiredResourceState();
        val foundAccelerator = HandlerCommons.getAccelerator(model.getAcceleratorArn(), proxy, agaClient, logger);
        if (foundAccelerator == null) {
            return ProgressEvent.defaultFailureHandler(
                    new Exception(String.format("Failed to find accelerator with arn: [%s].  Can not create listener", model.getAcceleratorArn())),
                    HandlerErrorCode.NotFound);
        }

        // if this is our first try then we will return
        if (model.getListenerArn() == null) {
            return createListenerStep(model, request, proxy, agaClient, logger);
        } else {
            return HandlerCommons.waitForSynchronizedStep(inferredCallbackContext, model, proxy, agaClient, logger);
        }
    }

    /**
     * Create an accelerator and create the correct progress continuation context
     */
    private ProgressEvent<ResourceModel, CallbackContext> createListenerStep(final ResourceModel model,
                                                                             final ResourceHandlerRequest<ResourceModel> handlerRequest,
                                                                             final AmazonWebServicesClientProxy proxy,
                                                                             final AWSGlobalAccelerator agaClient,
                                                                             final Logger logger) {
        logger.log("Creating new listener.");
        val listener = createListener(model, handlerRequest, proxy, agaClient);
        model.setListenerArn(listener.getListenerArn());
        model.setClientAffinity(listener.getClientAffinity());
        model.setProtocol(listener.getProtocol());

        val callbackContext = CallbackContext.builder()
                .stabilizationRetriesRemaining(HandlerCommons.NUMBER_OF_STATE_POLL_RETRIES)
                .build();

        return ProgressEvent.defaultInProgressHandler(callbackContext, 0, model);
    }

    private Listener createListener(final ResourceModel model,
                                    final ResourceHandlerRequest<ResourceModel> handlerRequest,
                                    final AmazonWebServicesClientProxy proxy,
                                    final AWSGlobalAccelerator agaClient) {
        val convertedPortRanges = model.getPortRanges().stream()
                .map(x -> new PortRange().withFromPort(x.getFromPort()).withToPort(x.getToPort()))
                .collect(Collectors.toList());

        val createListenerRequest = new CreateListenerRequest()
                .withAcceleratorArn(model.getAcceleratorArn())
                .withClientAffinity(model.getClientAffinity())
                .withProtocol(model.getProtocol())
                .withPortRanges(convertedPortRanges)
                .withIdempotencyToken(handlerRequest.getLogicalResourceIdentifier());

        return proxy.injectCredentialsAndInvoke(createListenerRequest, agaClient::createListener).getListener();
    }
}

File: aws-globalaccelerator-listener/src/main/kotlin/software/amazon/globalaccelerator/listener/CreateHandler.kt
package software.amazon.globalaccelerator.listener

import com.amazonaws.services.globalaccelerator.AWSGlobalAccelerator
import com.amazonaws.services.globalaccelerator.model.CreateListenerRequest
import com.amazonaws.services.globalaccelerator.model.Listener
import com.amazonaws.services.globalaccelerator.model.PortRange
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy
import software.amazon.cloudformation.proxy.HandlerErrorCode
import software.amazon.cloudformation.proxy.Logger
import software.amazon.cloudformation.proxy.ProgressEvent
import software.amazon.cloudformation.proxy.ResourceHandlerRequest

/**
 * Create handler implementation for listener resource.
 */
class CreateHandler : BaseHandler<CallbackContext>() {
    @Override
    override fun handleRequest(proxy: AmazonWebServicesClientProxy,
                               request: ResourceHandlerRequest<ResourceModel>,
                               callbackContext: CallbackContext?,
                               logger: Logger): ProgressEvent<ResourceModel, CallbackContext> {
        logger.debug("Create Listener Request [$request]")
        val agaClient = AcceleratorClientBuilder.client
        val model = request.desiredResourceState
        HandlerCommons.getAccelerator(model.acceleratorArn, proxy, agaClient, logger)
                ?: return ProgressEvent.defaultFailureHandler(
                        Exception("Failed to create listener. Cannot find accelerator with arn: [${model.acceleratorArn}]."),
                        HandlerErrorCode.NotFound)
        return createListenerStep(model, request, proxy, agaClient, logger)
    }

    /**
     * Create listener doesn't wait for accelerator to go in-sync
     */
    private fun createListenerStep(model: ResourceModel,
                                   handlerRequest: ResourceHandlerRequest<ResourceModel>,
                                   proxy: AmazonWebServicesClientProxy,
                                   agaClient: AWSGlobalAccelerator,
                                   logger: Logger): ProgressEvent<ResourceModel, CallbackContext> {
        logger.debug("Creating new listener with model: $model")
        val listener = createListener(model, handlerRequest, proxy, agaClient)
        model.apply {
            this.listenerArn = listener.listenerArn
            this.clientAffinity = listener.clientAffinity
            this.protocol = listener.protocol
        }
        return ProgressEvent.defaultSuccessHandler(model)
    }

    private fun createListener(model: ResourceModel,
                               handlerRequest: ResourceHandlerRequest<ResourceModel>,
                               proxy: AmazonWebServicesClientProxy,
                               agaClient: AWSGlobalAccelerator): Listener {
        val convertedPortRanges = model.portRanges.map { x -> PortRange().withFromPort(x.fromPort).withToPort(x.toPort) }
        val createListenerRequest = CreateListenerRequest()
                .withAcceleratorArn(model.acceleratorArn)
                .withClientAffinity(model.clientAffinity)
                .withProtocol(model.protocol)
                .withPortRanges(convertedPortRanges)
                .withIdempotencyToken(handlerRequest.clientRequestToken)
        return proxy.injectCredentialsAndInvoke(createListenerRequest, agaClient::createListener).listener
    }
}
